App backscaling client throws panic

During testing there was discovered and issue with client type casting:
```
{"level":"info","ts":1587406478.6430786,"logger":"discovery|vxspf","msg":"Pod updated.","meta":{"PK":"e13ff4f3af8482151b092d38134ab72527ed9da9","UID":"b310cb06-7abc-4c45-9393-9891a4c86bca","Version":1869477,"Namespace":"openshift-migration","Name":"registry-802065ca-2cc4-4f9b-aab3-714be6987fa4-9p269-1-deploy"}}
E0420 20:14:38.814093   84579 runtime.go:69] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x1a78b40), concrete:(*runtime._type)(0x1bcc600), asserted:(*runtime._type)(0x1c00b40), missingMethod:""} (interface conversion: client.Client is *compat.Client, not compat.Client)
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/lib/golang/src/runtime/panic.go:679
/usr/lib/golang/src/runtime/iface.go:255
/usr/lib/golang/src/runtime/iface.go:265
/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:52
/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:18
/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/validation.go:221
/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:215
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/lib/golang/src/runtime/asm_amd64.s:1357
panic: interface conversion: client.Client is *compat.Client, not compat.Client [recovered]
	panic: interface conversion: client.Client is *compat.Client, not compat.Client

goroutine 512 [running]:
github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x1a05420, 0xc006e83590)
	/usr/lib/golang/src/runtime/panic.go:679 +0x1b2
github.com/konveyor/mig-controller/pkg/controller/migplan.ReconcileMigPlan.newGVKCompare(0x1fc8120, 0xc0006ded50, 0xc00029eee0, 0xc00e99c1e0, 0xc00e99c300, 0x1c5218b, 0x9)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:52 +0x9cb
github.com/konveyor/mig-controller/pkg/controller/migplan.ReconcileMigPlan.compareGVK(0x1fc8120, 0xc0006ded50, 0xc00029eee0, 0xc00e99c1e0, 0x0, 0x0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/gvk.go:18 +0xdd
github.com/konveyor/mig-controller/pkg/controller/migplan.ReconcileMigPlan.validate(0x1fc8120, 0xc0006ded50, 0xc00029eee0, 0xc00e99c1e0, 0x0, 0xc00ed1794c)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/validation.go:221 +0x358
github.com/konveyor/mig-controller/pkg/controller/migplan.(*ReconcileMigPlan).Reconcile(0xc0005ff880, 0xc00ed81020, 0x13, 0xc00ed1794c, 0x3, 0xc001222600, 0x0, 0x0, 0x0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/pkg/controller/migplan/migplan_controller.go:215 +0x3a0
github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0005010e0, 0x1000000000000)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:215 +0x20a
github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1()
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:158 +0x36
github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc0074f24a0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x5e
github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0074f24a0, 0x3b9aca00, 0x0, 0x1, 0xc00004e1e0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xf8
github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc0074f24a0, 0x3b9aca00, 0xc00004e1e0)
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start
	/home/dgrigore/go/src/github.com/konveyor/mig-controller/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:157 +0x32e
exit status 2
make: *** [Makefile:21: run] Error 1
```

--- a/pkg/apis/migration/v1alpha1/migcluster_types.go
+++ b/pkg/apis/migration/v1alpha1/migcluster_types.go
@@ -18,12 +18,13 @@ package v1alpha1
 
 import (
 	"context"
+	"time"
+
 	pvdr "github.com/konveyor/mig-controller/pkg/cloudprovider"
 	"github.com/konveyor/mig-controller/pkg/compat"
 	"github.com/pkg/errors"
 	"k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/client/config"
-	"time"
 
 	ocapi "github.com/openshift/api/apps/v1"
 	imgapi "github.com/openshift/api/image/v1"
@@ -104,7 +105,7 @@ func (m *MigCluster) GetServiceAccountSecret(client k8sclient.Client) (*kapi.Sec
 }
 
 // GetClient get a local or remote client using a MigCluster and an existing client
-func (m *MigCluster) GetClient(c k8sclient.Client) (k8sclient.Client, error) {
+func (m *MigCluster) GetClient(c k8sclient.Client) (compat.Client, error) {
 	restConfig, err := m.BuildRestConfig(c)
 	if err != nil {
 		return nil, err

--- a/pkg/compat/client.go
+++ b/pkg/compat/client.go
@@ -20,7 +20,13 @@ import (
 //
 // A smart client.
 // Provides seamless API version compatibility.
-type Client struct {
+type Client interface {
+	k8sclient.Client
+	dapi.DiscoveryInterface
+	RestConfig() *rest.Config
+}
+
+type client struct {
 	*rest.Config
 	k8sclient.Client
 	dapi.DiscoveryInterface
@@ -32,7 +38,7 @@ type Client struct {
 
 //
 // Create a new client.
-func NewClient(restCfg *rest.Config) (k8sclient.Client, error) {
+func NewClient(restCfg *rest.Config) (Client, error) {
 	rClient, err := k8sclient.New(
 		restCfg,
 		k8sclient.Options{
@@ -57,7 +63,7 @@ func NewClient(restCfg *rest.Config) (k8sclient.Client, error) {
 	if err != nil {
 		return nil, err
 	}
-	nClient := &Client{
+	nClient := &client{
 		Config:             restCfg,
 		Client:             rClient,
 		DiscoveryInterface: dClient,
@@ -68,9 +74,13 @@ func NewClient(restCfg *rest.Config) (k8sclient.Client, error) {
 	return nClient, nil
 }
 
+func (c client) RestConfig() *rest.Config {
+	return c.Config
+}
+
 //
 // supportedVersion will determine correct version of the object provided, based on cluster version
-func (c Client) supportedVersion(obj runtime.Object) runtime.Object {
+func (c client) supportedVersion(obj runtime.Object) runtime.Object {
 	if c.Minor < 16 {
 		switch obj.(type) {
 
@@ -116,7 +126,7 @@ func (c Client) supportedVersion(obj runtime.Object) runtime.Object {
 
 //
 // Down convert a resource as needed based on cluster version.
-func (c Client) downConvert(ctx context.Context, obj runtime.Object) (runtime.Object, error) {
+func (c client) downConvert(ctx context.Context, obj runtime.Object) (runtime.Object, error) {
 	new := c.supportedVersion(obj)
 	if new == obj {
 		return obj, nil
@@ -132,7 +142,7 @@ func (c Client) downConvert(ctx context.Context, obj runtime.Object) (runtime.Ob
 
 //
 // upConvert will convert src resource to dst as needed based on cluster version
-func (c Client) upConvert(ctx context.Context, src runtime.Object, dst runtime.Object) error {
+func (c client) upConvert(ctx context.Context, src runtime.Object, dst runtime.Object) error {
 	if c.supportedVersion(dst) == dst {
 		dst = src
 		return nil
@@ -144,7 +154,7 @@ func (c Client) upConvert(ctx context.Context, src runtime.Object, dst runtime.O
 //
 // Get the specified resource.
 // The resource will be converted to a compatible version as needed.
-func (c Client) Get(ctx context.Context, key k8sclient.ObjectKey, in runtime.Object) error {
+func (c client) Get(ctx context.Context, key k8sclient.ObjectKey, in runtime.Object) error {
 	obj := c.supportedVersion(in)
 	err := c.Client.Get(ctx, key, obj)
 	if err != nil {
@@ -157,7 +167,7 @@ func (c Client) Get(ctx context.Context, key k8sclient.ObjectKey, in runtime.Obj
 //
 // List the specified resource.
 // The resource will be converted to a compatible version as needed.
-func (c Client) List(ctx context.Context, opt *k8sclient.ListOptions, in runtime.Object) error {
+func (c client) List(ctx context.Context, opt *k8sclient.ListOptions, in runtime.Object) error {
 	obj, err := c.downConvert(ctx, in)
 	if err != nil {
 		return err
@@ -172,7 +182,7 @@ func (c Client) List(ctx context.Context, opt *k8sclient.ListOptions, in runtime
 
 // Create the specified resource.
 // The resource will be converted to a compatible version as needed.
-func (c Client) Create(ctx context.Context, in runtime.Object) error {
+func (c client) Create(ctx context.Context, in runtime.Object) error {
 	obj, err := c.downConvert(ctx, in)
 	if err != nil {
 		return err
@@ -182,7 +192,7 @@ func (c Client) Create(ctx context.Context, in runtime.Object) error {
 
 // Delete the specified resource.
 // The resource will be converted to a compatible version as needed.
-func (c Client) Delete(ctx context.Context, in runtime.Object, opt ...k8sclient.DeleteOptionFunc) error {
+func (c client) Delete(ctx context.Context, in runtime.Object, opt ...k8sclient.DeleteOptionFunc) error {
 	obj, err := c.downConvert(ctx, in)
 	if err != nil {
 		return err
@@ -192,7 +202,7 @@ func (c Client) Delete(ctx context.Context, in runtime.Object, opt ...k8sclient.
 
 // Update the specified resource.
 // The resource will be converted to a compatible version as needed.
-func (c Client) Update(ctx context.Context, in runtime.Object) error {
+func (c client) Update(ctx context.Context, in runtime.Object) error {
 	obj, err := c.downConvert(ctx, in)
 	if err != nil {
 		return err

--- a/pkg/controller/migmigration/annotation.go
+++ b/pkg/controller/migmigration/annotation.go
@@ -5,6 +5,7 @@ import (
 	"strings"
 
 	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/konveyor/mig-controller/pkg/gvk"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	k8serror "k8s.io/apimachinery/pkg/api/errors"
@@ -490,7 +491,7 @@ func (t *Task) deleteNamespaceLabels(client k8sclient.Client, namespaceList []st
 // deleteLabels will delete all migration.openshift.io/migrated-by labels
 // from the application upon successful completion
 func (t *Task) deleteLabels() error {
-	client, GVRs, err := t.getResourcesForDelete()
+	client, GVRs, err := gvk.GetGVRsForCluster(t.PlanResources.DestMigCluster, t.Client)
 	if err != nil {
 		log.Trace(err)
 		return err

--- a/pkg/controller/migmigration/restore.go
+++ b/pkg/controller/migmigration/restore.go
@@ -6,14 +6,11 @@ import (
 	"strings"
 
 	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
-	"github.com/konveyor/mig-controller/pkg/compat"
 	"github.com/konveyor/mig-controller/pkg/gvk"
 	"github.com/pkg/errors"
 	velero "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
 	k8serror "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime/schema"
-	"k8s.io/client-go/dynamic"
 	"k8s.io/utils/pointer"
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
@@ -310,7 +307,7 @@ func (t *Task) deleteRestores() error {
 }
 
 func (t *Task) deleteMigrated() error {
-	client, GVRs, err := t.getResourcesForDelete()
+	client, GVRs, err := gvk.GetGVRsForCluster(t.PlanResources.DestMigCluster, t.Client)
 	if err != nil {
 		log.Trace(err)
 		return err
@@ -354,7 +351,7 @@ func (t *Task) deleteMigrated() error {
 }
 
 func (t *Task) ensureMigratedResourcesDeleted() (bool, error) {
-	client, GVRs, err := t.getResourcesForDelete()
+	client, GVRs, err := gvk.GetGVRsForCluster(t.PlanResources.DestMigCluster, t.Client)
 	if err != nil {
 		log.Trace(err)
 		return false, err
@@ -378,29 +375,4 @@ func (t *Task) ensureMigratedResourcesDeleted() (bool, error) {
 	}
 
 	return true, nil
-}
-
-func (t *Task) getResourcesForDelete() (dynamic.Interface, []schema.GroupVersionResource, error) {
-	c, err := t.getDestinationClient()
-	if err != nil {
-		log.Trace(err)
-		return nil, nil, err
-	}
-	dstClient := c.(*compat.Client)
-	client, err := dynamic.NewForConfig(dstClient.Config)
-	if err != nil {
-		log.Trace(err)
-		return nil, nil, err
-	}
-	resourceList, err := gvk.CollectResources(dstClient)
-	if err != nil {
-		log.Trace(err)
-		return nil, nil, err
-	}
-	GVRs, err := gvk.ConvertToGVRList(resourceList)
-	if err != nil {
-		log.Trace(err)
-		return nil, nil, err
-	}
-	return client, GVRs, nil
 }

--- a/pkg/controller/migmigration/task.go
+++ b/pkg/controller/migmigration/task.go
@@ -5,6 +5,7 @@ import (
 
 	"github.com/go-logr/logr"
 	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
+	"github.com/konveyor/mig-controller/pkg/compat"
 	"github.com/pkg/errors"
 	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
@@ -815,12 +816,12 @@ func (t *Task) keepAnnotations() bool {
 }
 
 // Get a client for the source cluster.
-func (t *Task) getSourceClient() (k8sclient.Client, error) {
+func (t *Task) getSourceClient() (compat.Client, error) {
 	return t.PlanResources.SrcMigCluster.GetClient(t.Client)
 }
 
 // Get a client for the destination cluster.
-func (t *Task) getDestinationClient() (k8sclient.Client, error) {
+func (t *Task) getDestinationClient() (compat.Client, error) {
 	return t.PlanResources.DestMigCluster.GetClient(t.Client)
 }
 

--- a/pkg/controller/migplan/gvk.go
+++ b/pkg/controller/migplan/gvk.go
@@ -2,7 +2,6 @@ package migplan
 
 import (
 	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
-	"github.com/konveyor/mig-controller/pkg/compat"
 	"github.com/konveyor/mig-controller/pkg/gvk"
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/dynamic"
@@ -44,19 +43,17 @@ func (r ReconcileMigPlan) newGVKCompare(plan *migapi.MigPlan) (*gvk.Compare, err
 		return nil, err
 	}
 
-	src, err := srcCluster.GetClient(r)
+	srcClient, err := srcCluster.GetClient(r)
 	if err != nil {
 		log.Trace(err)
 		return nil, err
 	}
-	srcClient := src.(compat.Client)
-	dst, err := dstCluster.GetClient(r)
+	dstClient, err := dstCluster.GetClient(r)
 	if err != nil {
 		log.Trace(err)
 		return nil, err
 	}
-	dstClient := dst.(compat.Client)
-	dynamicClient, err := dynamic.NewForConfig(srcClient.Config)
+	dynamicClient, err := dynamic.NewForConfig(srcClient.RestConfig())
 	if err != nil {
 		log.Trace(err)
 		return nil, err

--- a/pkg/gvk/gvk.go
+++ b/pkg/gvk/gvk.go
@@ -7,6 +7,7 @@ import (
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"k8s.io/client-go/discovery"
 	"k8s.io/client-go/dynamic"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 
 	migapi "github.com/konveyor/mig-controller/pkg/apis/migration/v1alpha1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -29,12 +30,12 @@ type Compare struct {
 // Compare GVKs on both clusters, find incompatible GVKs
 // and check each plan source namespace for existence of incompatible GVKs
 func (r *Compare) Compare() (map[string][]schema.GroupVersionResource, error) {
-	srcResourceList, err := CollectResources(r.SrcDiscovery)
+	srcResourceList, err := collectResources(r.SrcDiscovery)
 	if err != nil {
 		return nil, err
 	}
 
-	dstResourceList, err := CollectResources(r.DstDiscovery)
+	dstResourceList, err := collectResources(r.DstDiscovery)
 	if err != nil {
 		return nil, err
 	}
@@ -45,7 +46,7 @@ func (r *Compare) Compare() (map[string][]schema.GroupVersionResource, error) {
 	}
 
 	resourcesDiff := compareResources(srcResourceList, dstResourceList)
-	incompatibleGVKs, err := ConvertToGVRList(resourcesDiff)
+	incompatibleGVKs, err := convertToGVRList(resourcesDiff)
 	if err != nil {
 		return nil, err
 	}
@@ -54,7 +55,7 @@ func (r *Compare) Compare() (map[string][]schema.GroupVersionResource, error) {
 }
 
 // CollectResources collects all namespaced scoped apiResources from the cluster
-func CollectResources(discovery discovery.DiscoveryInterface) ([]*metav1.APIResourceList, error) {
+func collectResources(discovery discovery.DiscoveryInterface) ([]*metav1.APIResourceList, error) {
 	resources, err := discovery.ServerResources()
 	if err != nil {
 		return nil, err
@@ -70,8 +71,8 @@ func CollectResources(discovery discovery.DiscoveryInterface) ([]*metav1.APIReso
 	return resources, nil
 }
 
-// ConvertToGVRList converts provided apiResourceList to list of GroupVersionResources from the server
-func ConvertToGVRList(resourceList []*metav1.APIResourceList) ([]schema.GroupVersionResource, error) {
+// convertToGVRList converts provided apiResourceList to list of GroupVersionResources from the server
+func convertToGVRList(resourceList []*metav1.APIResourceList) ([]schema.GroupVersionResource, error) {
 	GVRs := []schema.GroupVersionResource{}
 	for _, resourceList := range resourceList {
 		gv, err := schema.ParseGroupVersion(resourceList.GroupVersion)
@@ -86,6 +87,27 @@ func ConvertToGVRList(resourceList []*metav1.APIResourceList) ([]schema.GroupVer
 	}
 
 	return GVRs, nil
+}
+
+// GetGVRsForCluster collects all namespaced scoped GVRs for the provided cluster compatible client
+func GetGVRsForCluster(cluster *migapi.MigCluster, c client.Client) (dynamic.Interface, []schema.GroupVersionResource, error) {
+	compat, err := cluster.GetClient(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	dynamic, err := dynamic.NewForConfig(compat.RestConfig())
+	if err != nil {
+		return nil, nil, err
+	}
+	resourceList, err := collectResources(compat)
+	if err != nil {
+		return nil, nil, err
+	}
+	GVRs, err := convertToGVRList(resourceList)
+	if err != nil {
+		return nil, nil, err
+	}
+	return dynamic, GVRs, nil
 }
 
 func excludeSubresources(resources []metav1.APIResource) []metav1.APIResource {

